### PR TITLE
Add compute items for spacetime Christoffels

### PR DESCRIPTION
**Pull request by prayush**
## Proposed changes

Add compute items for spacetime Christoffels for first and second kind, and the trace of the first.

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: src/PointwiseFunctions/GeneralRelativity/Christoffel.hpp [hunk 1]
```diff
@@ -38,4 +45,85 @@ template <size_t SpatialDim, typename Frame, IndexType Index, typename DataType>
...
+      &raise_or_lower_first_index<DataType,
+                                  SpacetimeIndex<SpatialDim, UpLo::Lo, Frame>,
+                                  SpacetimeIndex<SpatialDim, UpLo::Lo, Frame>>;
+  using base = SpacetimeChristoffelSecondKind<SpatialDim, Frame, DataType>;
+  using type = typename base::type;
+};
+
+/// Compute item for the trace of the spacetime Christoffel symbols
+/// of the first kind
+/// \f$\Gamma_{a} = \Gamma_{abc}g^{bc}\f$ compputed from the
```
Comment (carmaza) [inline src/PointwiseFunctions/GeneralRelativity/Christoffel.hpp]: computed

**Review by carmaza — COMMENTED** [review]
You can squash this fixup too.

**Review by carmaza — DISMISSED** [review]

File: src/PointwiseFunctions/GeneralRelativity/Christoffel.hpp [hunk 2]
```diff
@@ -38,4 +45,85 @@ template <size_t SpatialDim, typename Frame, IndexType Index, typename DataType>
...
+template <size_t SpatialDim, typename Frame, typename DataType>
+struct SpacetimeChristoffelSecondKindCompute
+    : SpacetimeChristoffelSecondKind<SpatialDim, Frame, DataType>,
+      db::ComputeTag {
+  using argument_tags =
+      tmpl::list<SpacetimeChristoffelFirstKind<SpatialDim, Frame, DataType>,
+                 InverseSpacetimeMetric<SpatialDim, Frame, DataType>>;
+  static constexpr Tensor<
+      DataType, Symmetry<2, 1, 1>,
+      index_list<SpacetimeIndex<SpatialDim, UpLo::Up, Frame>,
```
Comment (moxcodes) [inline src/PointwiseFunctions/GeneralRelativity/Christoffel.hpp]: The `SpacetimeIndex` structs are used a bunch in these function pointer declarations. This could be made more brief with a couple of type aliases for them, like `using up_index = SpacetimeIndex<SpatialDim, UpLo::Up, Frame>`. I won't insist on this change; I'll leave it up to you whether you feel it improves the struct.

File: tests/Unit/PointwiseFunctions/GeneralRelativity/Test_Christoffel.cpp [hunk 3]
```diff
@@ -39,4 +54,136 @@ SPECTRE_TEST_CASE("Unit.PointwiseFunctions.GeneralRelativity.Christoffel.",
...
+  CHECK(gr::Tags::SpacetimeChristoffelSecondKindCompute<3, Frame::Inertial,
+                                                        DataVector>::name() ==
+        "SpacetimeChristoffelSecondKind");
+
+  // Check that the compute items return correct values
+  const DataVector test_vector{5., 4.};
+  auto spacetime_metric =
+      make_with_value<tnsr::aa<DataVector, 3, Frame::Inertial>>(test_vector,
+                                                                0.);
+  get<0, 0>(spacetime_metric) = -1.5;
```
Comment (moxcodes) [inline tests/Unit/PointwiseFunctions/GeneralRelativity/Test_Christoffel.cpp]: This might be a good place to use `make_with_random_values` or `fill_with_random_values`, similar to my suggestion in #1478 .

File: src/PointwiseFunctions/GeneralRelativity/Christoffel.hpp (hunk 2 above)
Comment (wthrowe) [reply src/PointwiseFunctions/GeneralRelativity/Christoffel.hpp]: Isn't this whole thing just `tnsr::Abb<DataType, SpatialDim>`?

File: tests/Unit/PointwiseFunctions/GeneralRelativity/Test_Christoffel.cpp (hunk 3 above)
Comment (prayush) [reply tests/Unit/PointwiseFunctions/GeneralRelativity/Test_Christoffel.cpp]: changed to using `make_with_random_values` as in #1478

File: src/PointwiseFunctions/GeneralRelativity/Christoffel.hpp (hunk 2 above)
Comment (prayush) [reply src/PointwiseFunctions/GeneralRelativity/Christoffel.hpp]: replaced these long expressions with abbreviated `tnsr` aliases now.

Comment (moxcodes) [reply src/PointwiseFunctions/GeneralRelativity/Christoffel.hpp]: oh, yes, the abbreviated name is a significant improvement. :+1:

File: tests/Unit/PointwiseFunctions/GeneralRelativity/Test_Christoffel.cpp [hunk 4]
```diff
@@ -39,4 +56,93 @@ SPECTRE_TEST_CASE("Unit.PointwiseFunctions.GeneralRelativity.Christoffel.",
...
+        "SpacetimeChristoffelSecondKind");
+
+  // Check that the compute items return correct values
+  const DataVector test_vector{5., 4.};
+  MAKE_GENERATOR(generator);
+  std::uniform_real_distribution<> distribution(-0.1, 0.1);
+
+  auto spacetime_metric =
+      make_with_random_values<tnsr::aa<DataVector, 3, Frame::Inertial>>(
+          make_not_null(&generator), make_not_null(&distribution), test_vector);
```
Comment (moxcodes) [inline tests/Unit/PointwiseFunctions/GeneralRelativity/Test_Christoffel.cpp]: Same note as in #1478: if the `test_vector` is just used for size, then you can probably just input the numeric size you want instead.

**Review by moxcodes — COMMENTED** [review]
I'm happy with squashing the changes, including any alteration to the `test_vector` you decide to make.

File: tests/Unit/PointwiseFunctions/GeneralRelativity/Test_Christoffel.cpp (hunk 4 above)
Comment (prayush) [reply tests/Unit/PointwiseFunctions/GeneralRelativity/Test_Christoffel.cpp]: squashed

**Review by moxcodes — DISMISSED** [review]

File: src/PointwiseFunctions/GeneralRelativity/Tags.hpp [hunk 5]
```diff
@@ -47,7 +47,11 @@ struct Lapse : db::SimpleTag {
   using type = Scalar<DataType>;
   static std::string name() noexcept { return "Lapse"; }
 };
-
+template <size_t Dim, typename Frame, typename DataType>
+struct DerivativesOfSpacetimeMetric : db::SimpleTag {
```
Comment (nilsdeppe) [inline src/PointwiseFunctions/GeneralRelativity/Tags.hpp:66]: Could you please add a short doxygen comment so this at least shows up in the documentation? Probably most of the tags in this file need that, but I'm fine with that being done in a separate PR. I've notated some of the classes with no dox in #1514

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
I have one small request, please squash and rebase on develop right away :)

File: src/PointwiseFunctions/GeneralRelativity/Tags.hpp (hunk 5 above)
Comment (prayush) [reply src/PointwiseFunctions/GeneralRelativity/Tags.hpp:66]: added, squashed and rebased

**Review by geoffrey4444 — DISMISSED** [review]

**Review by moxcodes — DISMISSED** [review]

**Comment by nilsdeppe** [comment]
@prayush needs to be rebased. This has conflicts now apparently with the other PR I just merged.

**Comment by prayush** [comment]
> @prayush needs to be rebased. This has conflicts now apparently with the other PR I just merged.

@nilsdeppe - rebased :)

**Review by moxcodes — APPROVED** [review]

**Review by nilsdeppe — APPROVED** [review]

**Comment by nilsdeppe** [comment]
GCC6 debug timed out before completing tests. Restarted both stage 1 and stage 2.

**Comment by nilsdeppe** [comment]
GCC6 debug timed out again but all tests passed.
